Fix date parser literals

vmlf01 opened the pull request:
This will be a fix for #1897.

However, I'm having some trouble coming up with a regular expression to cope with escaped quotes inside the literals, i.e. ```3 '0''clock'```, so it's still not finished.

Help anyone?

mgcrea commented:
You mean `3 '0\'clock'`? Or actually `3 '0''clock'`? Is a double single quote an escape?

mgcrea commented:
[This patch](https://github.com/mgcrea/angular-strap/commit/7a1fd530daeed410e6c462472ce2bd9c7f9fbe12) seems to pass the tests, what do you think?

vmlf01 commented:
I mean the second case, from angular date filter (https://github.com/angular/angular.js/blob/master/src/ng/filter/filters.js#L402), they are using two single quotes as an escape sequence, instead of a back-slash.

But your patch gave me another idea, maybe we can replace two consecutive single quotes with the back-slash escape sequence at the beginning and then use the regex for escaped quotes to match it.
That might be easier to understand than handle the empty formatPart tokens.

vmlf01 commented:
Changed last commit to support the escaped literals, but now it seems that the date value parsing in https://github.com/mgcrea/angular-strap/blob/master/src/helpers/date-parser.js#L231 also has the original problem.

It seems very similar to the regex build code just fixed, so maybe it is possible to re-use it. I'll continue to check tomorrow.

vmlf01 commented:
I think I fixed everything now, but I still have a question about line https://github.com/vmlf01/angular-strap/blob/77cb19cf48735f717334e7c1b27ce6a81d4e9f50/src/helpers/date-parser.js#L281

Before it was
```
        format = escapeReservedSymbols(format);
```
and I changed it to 
```
        re = escapeReservedSymbols(re);
```
because ```format``` is a local variable, so that line did nothing, and I thought it was just a typing error, but in reality, I don't know what the escapeReservedSymbols purpose is or if it is needed.

mgcrea commented:
Hum. Looks like a good fix, it looks like `escapeReservedSymbols` is supposed to escape symbols that have special meaning in a regex, for instance, a matches any character, but if you want to specifically match a dot, you have to `replace(/\./g, '[.]')`.

mgcrea commented:
Hum having a second look at it, I don't think it's suppose to be used there, as it will break matched patterns. It should probably be applied before on the format, we need extra tests to check this.

belfz commented:
Has this been released yet? I think that the problem referenced in #1897 still occurs even with version 2.3.5

vmlf01 commented:
No it hasn't, it will be in next version.
